tests/synthdnn/matmul/sampler.py

```diff
@@ -0,0 +1,148 @@
+        else:
+            raise RuntimeError(f"Unknown iteration mode {self.mode}")
+
+
+    # Itertools.product seems to break on an infinite sampler
+    class ProductIter:
+       def __init__(self, samples, kinds, dim_sampler):
+           self.dim_sampler = dim_sampler
+           self.kinds = kinds
+           self.kinds_iter = iter(self.kinds)
+           self.rem_samples = samples
+
+       def __next__(self):
+           if(self.rem_samples == 0):
+               raise StopIteration
+
+           try:
+               self.k = next(self.kinds_iter);
+               self.s = next(self.dim_sampler)
+           except StopIteration:
+               self.kinds_iter = iter(self.kinds)
+               self.k = next(self.kinds_iter)
+               self.s = next(self.dim_sampler)
+               self.rem_samples = self.rem_samples - 1
+
+           return Primitive(self.k, self.s)
+
+    class ZipIter:
+       def __init__(self, samples, kinds, dim_sampler):
+           self.dim_sampler = dim_sampler
+           self.kinds_iter = itertools.cycle(kinds)
+           self.rem_samples = samples
+
+       def __next__(self):
+           if(self.rem_samples == 0):
+               raise StopIteration
+
+           self.rem_samples = self.rem_samples - 1
+           k = next(self.kinds_iter);
+           s = next(self.dim_sampler)
+
+           return Primitive(k, s)
+
+    class DimSampler:
+        def __init__(self, region):
+            self.region = region
+            self.seen = set()
+            if region.ndims < 3:
+               raise RuntimeError(f"Insufficient dimensions for matmul operation, expected at least 3, but got {region.ndims}")
+
+        def __next__(self):
+
+            # Sample from a power distribution as most problem features occur
+            # when some dimension is small. In addition, small problems often
+            # require less time to run enabling faster data collection
+            def get_sample(minval, maxval, align):
+                assert minval <= maxval, "Sample bounds are out of order"
+                if minval == maxval:
+                    return minval
+                x = round(pow(2, random.uniform(math.log2(minval), math.log2(maxval))))
+                return (x // align) * align
+
+            for _ in range(1000):
+                dims = [0] * self.region.ndims
+                for i in range(self.region.ndims):
+                    dims[i] = get_sample(
+                        self.region.min[i],
+                        self.region.max[i],
+                        self.region.alignment[i],
+                        )
+                dims_tuple = tuple(dims)
+                if dims_tuple not in self.seen:
+                    self.seen.add(dims_tuple)
+                    return Dims(dims[:-3], dims[-3], dims[-2], dims[-1])
+
+            raise RuntimeError(
+                "Can not effectively sample problems larger than: "
+                + str(min_size)
```

Comment:
This variable doesn't exist. Should probably be something like:
```suggestion
                f"Can not sample >{len(self.seen)} problems in region {self.region}"
```